File: android/share/src/main/java/com/dreampany/share/vm/ApkViewModel.java
package com.dreampany.share.vm;

import android.app.Application;

import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;

import com.dreampany.framework.data.model.Response;
import com.dreampany.framework.misc.AppExecutors;
import com.dreampany.framework.misc.ResponseMapper;
import com.dreampany.framework.misc.RxMapper;
import com.dreampany.framework.misc.SmartMap;
import com.dreampany.framework.misc.exceptions.EmptyException;
import com.dreampany.framework.misc.exceptions.ExtraException;
import com.dreampany.framework.misc.exceptions.MultiException;
import com.dreampany.framework.ui.vm.BaseViewModel;
import com.dreampany.media.data.enums.MediaType;
import com.dreampany.media.data.model.Apk;
import com.dreampany.media.data.source.repository.ApkRepository;
import com.dreampany.share.data.model.SelectEvent;
import com.dreampany.share.data.source.repository.share.ApkShareRepository;
import com.dreampany.share.misc.Comparators;
import com.dreampany.share.misc.exception.SelectException;
import com.dreampany.share.ui.model.MediaItem;
import com.dreampany.share.ui.model.UiTask;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import javax.inject.Inject;


import io.reactivex.Flowable;
import io.reactivex.Maybe;
import io.reactivex.MaybeSource;
import io.reactivex.Single;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;
import timber.log.Timber;

/**
 * Created by Hawladar Roman on 7/18/2018.
 * BJIT Group
 * hawladar.roman@bjitgroup.com
 */

public class ApkViewModel extends BaseViewModel<Apk, MediaItem, UiTask<Apk>> {

    private final ApkRepository repo;
    private final ApkShareRepository shareRepo;
    private final Comparators comparators;

    @Inject
    ApkViewModel(Application application,
                 RxMapper rx,
                 AppExecutors ex,
                 ResponseMapper rm,
                 ApkRepository repo,
                 ApkShareRepository shareRepo,
                 Comparators comparators) {
        super(application, rx, ex, rm);
        this.repo = repo;
        this.shareRepo = shareRepo;
        this.comparators = comparators;
    }

    @Override
    public void clear() {
        super.clear();
    }


    public void loads(boolean important, boolean progress) {
        if (!takeAction(important, getMultipleDisposable())) {
            return;
        }
        Disposable disposable = getRx()
                .backToMain(getItemsRx())
                .doOnSubscribe(subscription -> {
                    if (progress) {
                        postProgress(true);
                    }
                })
                .subscribe(result -> {
                    if (progress) {
                        postProgress(false);
                    }
                    postResult(Response.Type.GET, result);
                }, error -> {
                    if (progress) {
                        postProgress(false);
                    }
                    postFailures(new MultiException(error, new ExtraException()));
                });
        addMultipleSubscription(disposable);
    }


    public void loadsWithShare(boolean important, boolean progress) {
        getEx().postToUi(() -> {
            if (!takeAction(important, getMultipleDisposable())) {
                return;
            }
            Disposable disposable = getRx()
                    .backToMain(getItemsWithShareRx())
                    .doOnSubscribe(subscription -> {
                        if (progress) {
                            postProgress(true);
                        }
                    })
                    .subscribe(result -> {
                        if (progress) {
                            postProgress(false);
                        }
                        postResult(Response.Type.GET, result);
                        notifySelect();
                    }, error -> {
                        if (progress) {
                            postProgress(false);
                        }
                        postFailures(new MultiException(error, new ExtraException()));
                    });
            addMultipleSubscription(disposable);
        }, 500L);
    }


    public void loadsShared(boolean important, boolean progress) {
        if (!takeAction(important, getMultipleDisposable())) {
            return;
        }
        Disposable disposable = getRx()
                .backToMain(getSharedItemsRx())
                .doOnSubscribe(subscription -> {
                    if (progress) {
                        postProgress(true);
                    }
                })
                .subscribe(result -> {
                    if (progress) {
                        postProgress(false);
                    }
                    postResult(Response.Type.GET, result);
                }, error -> {
                    if (progress) {
                        postProgress(false);
                    }
                    postFailures(new MultiException(error, new ExtraException()));
                });
        addMultipleSubscription(disposable);
    }

    public void toggleSelect(Apk apk) {
        if (hasSingleDisposable()) {
            return;
        }
        Disposable disposable = getRx()
                .backToMain(toggleSelectImpl(apk))
                .subscribe(item -> {
                    postResult(Response.Type.UPDATE, item);
                    notifySelect();
                }, this::postFailure);
        addSingleSubscription(disposable);
    }

    public void toggleShare(Apk item) {

    }

    public void toggleShare(List<Apk> items) {

    }

    public void selectToShare() {
        if (!takeAction(false, getMultipleDisposable())) {
            return;
        }
        Disposable disposable = getRx()
                .backToMain(putSharedItemsRx())
                .doOnSubscribe(subscription -> {
                    postProgress(true);
                }).subscribe(items -> {
                    getUiSelects().clear();
                    notifySelect();
                }, this::postFailure);
        addMultipleSubscription(disposable);
    }


    public void notifySelect() {
        if (!takeAction(false, getSingleDisposable())) {
            return;
        }
        Disposable disposable = getRx()
                .backToMain(getSelectEventRx())
                .subscribe(new Consumer<SelectEvent>() {
                    @Override
                    public void accept(SelectEvent selectEvent) throws Exception {

                    }
                }, this::postFailure);
        addSingleSubscription(disposable);
    }

    private Maybe<List<MediaItem>> getItemsRx() {
        return repo.getItemsRx()
                .flatMap((Function<List<Apk>, MaybeSource<List<MediaItem>>>) this::getItemsRx);
    }

    private Maybe<List<MediaItem>> getItemsWithShareRx() {
        return repo.getItemsRx()
                .flatMap((Function<List<Apk>, MaybeSource<List<MediaItem>>>) this::getItemsWithShareRx);
    }

    private Maybe<List<MediaItem>> getSharedItemsRx() {
        return shareRepo.getSharedItemsRx(repo)
                .flatMap((Function<List<Apk>, MaybeSource<List<MediaItem>>>) this::getSharedItemsRx);
    }

    private Maybe<List<Long>> putSharedItemsRx() {
        if (!hasSelection()) {
            return Maybe.error(new EmptyException());
        }
        List<Apk> items = new ArrayList<>(getUiSelects());
        return shareRepo.putItemsRx(items);
    }


    private Maybe<List<MediaItem>> getItemsRx(List<Apk> result) {
        return Maybe.fromCallable(() -> {
            List<MediaItem> items = new ArrayList<>(result.size());

            for (Apk apk : result) {
                MediaItem item = getItem(apk);
                items.add(item);
            }
            return items;
        });
    }


    private Maybe<List<MediaItem>> getItemsWithShareRx(List<Apk> result) {
        return Maybe.fromCallable(() -> {
            Collections.sort(result, comparators.getDisplayNameComparator());
            List<MediaItem> items = new ArrayList<>(result.size());
            for (Apk apk : result) {
                MediaItem item = getItemWithShare(apk);
                items.add(item);
            }
            return items;
        });
    }


    private Maybe<List<MediaItem>> getSharedItemsRx(List<Apk> result) {
        return Maybe.fromCallable(() -> {
            List<MediaItem> items = new ArrayList<>(result.size());

            for (Apk apk : result) {
                MediaItem item = getItem(apk);
                items.add(item);
            }
            return items;
        });
    }

    private Single<SelectEvent> getSelectEventRx() {
        return Single.fromCallable(() -> {
            int selected = getUiSelects().size();
            int total = repo.getCount();
            SelectEvent event = new SelectEvent(MediaType.APK, selected, total);
            return event;
        });
    }

    private MediaItem getItem(Apk apk) {
        SmartMap<Long, MediaItem> map = getUiMap();
        MediaItem item = map.get(apk.getId());
        if (item == null) {
            item = MediaItem.getItem(apk);
            map.put(apk.getId(), item);
        }
        return item;
    }

    private MediaItem getItemWithShare(Apk apk) {
        SmartMap<Long, MediaItem> map = getUiMap();
        MediaItem item = map.get(apk.getId());
        if (item == null) {
            item = MediaItem.getItem(apk);
            map.put(apk.getId(), item);
        }
        adjustShare(item);
        adjustSelect(item);
        return item;
    }

    private void adjustShare(MediaItem item) {
        boolean shared = shareRepo.isShared((Apk) item.getItem());
        item.setShared(shared);
    }

    private void adjustSelect(MediaItem item) {
        boolean selected = getUiSelects().contains((Apk) item.getItem());
        item.setSelected(selected);
    }

    private Flowable<MediaItem> toggleSelectImpl(Apk apk) {
        return Flowable.fromCallable(() -> {
            boolean shared = shareRepo.isShared(apk);
            if (shared) {
                throw new SelectException("This item is already shared!");
            }
            if (getUiSelects().contains(apk)) {
                getUiSelects().remove(apk);
            } else {
                getUiSelects().add(apk);
            }
            MediaItem item = getItemWithShare(apk);
            return item;
        });
    }
}

File: android/nearby/src/main/java/com/dreampany/nearby/ui/publish/vm/ApkViewModel.kt
package com.dreampany.nearby.ui.publish.vm

import android.app.Application
import com.dreampany.framework.misc.func.ResponseMapper
import com.dreampany.framework.misc.func.SmartError
import com.dreampany.framework.ui.model.UiTask
import com.dreampany.framework.ui.vm.BaseViewModel
import com.dreampany.nearby.data.enums.Action
import com.dreampany.nearby.data.enums.State
import com.dreampany.nearby.data.enums.Subtype
import com.dreampany.nearby.data.enums.Type
import com.dreampany.nearby.data.model.media.Apk
import com.dreampany.nearby.data.source.pref.AppPref
import com.dreampany.nearby.data.source.repo.ApkRepo
import com.dreampany.nearby.ui.publish.model.ApkItem
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import timber.log.Timber
import javax.inject.Inject

/**
 * Created by roman on 29/6/20
 * Copyright (c) 2020 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
class ApkViewModel
@Inject constructor(
    application: Application,
    rm: ResponseMapper,
    private val pref: AppPref,
    private val repo: ApkRepo
) : BaseViewModel<Type, Subtype, State, Action, Apk, ApkItem, UiTask<Type, Subtype, State, Action, Apk>>(
    application,
    rm
) {

    fun loadApks() {
        uiScope.launch {
            postProgressMultiple(true)
            var result: List<Apk>? = null
            var errors: SmartError? = null
            try {
                result = repo.gets()
            } catch (error: SmartError) {
                Timber.e(error)
                errors = error
            }
            if (errors != null) {
                postError(errors)
            } else {
                postResult(result?.toItems())
            }
        }
    }

    fun loadApk(id: String) {
        uiScope.launch {
            postProgressSingle(true)
            var result: Apk? = null
            var errors: SmartError? = null
            var favorite: Boolean = false
            try {
                result = repo.get(id)
                result?.let { favorite = repo.isFavorite(it) }
            } catch (error: SmartError) {
                Timber.e(error)
                errors = error
            }
            if (errors != null) {
                postError(errors)
            } else {
                postResult(result?.toItem(favorite))
            }
        }
    }

    fun loadFavoriteApks() {
        uiScope.launch {
            postProgressMultiple(true)
            var result: List<Apk>? = null
            var errors: SmartError? = null
            try {
                result = repo.getFavorites()
            } catch (error: SmartError) {
                Timber.e(error)
                errors = error
            }
            if (errors != null) {
                postError(errors)
            } else {
                postResult(result?.toItems())
            }
        }
    }

    fun toggleFavorite(input: Apk) {
        uiScope.launch {
            postProgressSingle(true)
            var result: Apk? = null
            var errors: SmartError? = null
            var favorite: Boolean = false
            try {
                favorite = repo.toggleFavorite(input)
                result = input
            } catch (error: SmartError) {
                Timber.e(error)
                errors = error
            }
            if (errors != null) {
                postError(errors)
            } else {
                postResult(result?.toItem(favorite), state = State.FAVORITE)
            }
        }
    }

    private suspend fun List<Apk>.toItems(): List<ApkItem> {
        val input = this
        return withContext(Dispatchers.IO) {
            input.map { input ->
                val favorite = repo.isFavorite(input)
                ApkItem(input, favorite)
            }
        }
    }

    private suspend fun Apk.toItem(favorite: Boolean): ApkItem {
        val input = this
        return withContext(Dispatchers.IO) {
            ApkItem(input, favorite = favorite)
        }
    }

    private fun postProgressSingle(progress: Boolean) {
        postProgressSingle(
            Type.APK,
            Subtype.DEFAULT,
            State.DEFAULT,
            Action.DEFAULT,
            progress = progress
        )
    }

    private fun postProgressMultiple(progress: Boolean) {
        postProgressMultiple(
            Type.APK,
            Subtype.DEFAULT,
            State.DEFAULT,
            Action.DEFAULT,
            progress = progress
        )
    }


    private fun postError(error: SmartError) {
        postMultiple(
            Type.APK,
            Subtype.DEFAULT,
            State.DEFAULT,
            Action.DEFAULT,
            error = error,
            showProgress = true
        )
    }

    private fun postResult(result: List<ApkItem>?) {
        postMultiple(
            Type.APK,
            Subtype.DEFAULT,
            State.DEFAULT,
            Action.DEFAULT,
            result = result,
            showProgress = true
        )
    }

    private fun postResult(result: ApkItem?, state: State = State.DEFAULT) {
        postSingle(
            Type.APK,
            Subtype.DEFAULT,
            state,
            Action.DEFAULT,
            result = result,
            showProgress = true
        )
    }
}